Properly deregister a particle if deleted during refinement

gassmoeller opened the pull request:
During mesh refinement with particles, we search all children of a parent cell for the child that contains the particle. Sometimes we do not find the particle in any child, in which case we delete the particle. Presumably for performance reasons we do not call the `remove_particle` function (which repeats a bunch of operations we already did here), but we forget to properly de-register the particle handle with the property pool, which means the particle memory is never returned. This is not terrible, as the memory leak will be fixed the next time  `sort_particles_into_subdomains_and_cells` is called and only the memory of still existing particles is kept. However this bug can trigger an Assert if we call `PropertyPool<dim, spacedim>::clear()` before doing that, because the PropertyPool notices that there are dangling handles that have not been de-registered (and of course it leads to an inconsistent state of the property pool as well). This PR should fix the issue and make some minor improvements to the code, but I still need to make a test and changelog entry. I opened this early to let @ryanstoner1 and @vaturino who reported this first in https://github.com/geodynamics/aspect/issues/5594 test if this change fixes their issues (it did for my test model).

bangerth reviewed (approved):
Nice work tracking this down!

gassmoeller commented:
Alright, I got the confirmation in geodynamics/aspect#5594 that this PR fixes the reported issue. I added a changelog entry and test case. While adding the test case I noticed we already have a test that should have noticed this bug (`particles/particle_handler_refinement_outside_cell`). However, the test didnt trigger, because the property pool only did the memory consistency check if particles would carry additional properties. This makes no sense, and I think it is a leftover from the time when the property pool only stored particle properties (not locations). I tightened the memory check in the property pool to always check for consistent handles, and also added a second test that uses particles with additional properties. Both tests fail on master, but pass with this branch. This PR is ready for review from my side.

bangerth commented on source/particles/property_pool.cc:
```diff
@@ -47,19 +47,16 @@ namespace Particles
   void
   PropertyPool<dim, spacedim>::clear()
   {
-    if (n_properties > 0)
```
Ha, yes, funny, isn't it?
